ts/query.go

```diff
@@ -0,0 +1,438 @@
+	var sum float64
+	for i := range is {
+		sum += is[i].avg()
+	}
+	return sum
+}
+
+// Query returns data for the named time series during the supplied time span.
+// Data is returned as a series of consecutive data points.
+//
+// Data is queried only at the Resolution supplied: if data for the named time
+// series is not stored at the given resolution, an empty result will be
+// returned.
+//
+// All data stored on the server is downsampled to some degree; the data points
+// returned represent the average value within a sample period. Each datapoint's
+// timestamp falls in the middle of the sample period it represents.
+//
+// If data for a time series was collected from multiple sources, datapoints
+// will represent the sum across all sources at the same sample period. The
+// response will contain a list of all sources which were summed to produce the
+// result.
+func (db *DB) Query(name string, r Resolution, startNanos, endNanos int64) (
+	*proto.TimeSeriesQueryResult, error) {
+	// Normalize startNanos and endNanos the nearest SampleDuration boundary.
+	startNanos -= startNanos % r.SampleDuration()
+
+	// Based on the supplied timestamps and resolution, construct start and end
+	// keys for a scan that will return every key with data relevant to the
+	// query.
+	startKey := MakeDataKey(name, "" /* source */, r, startNanos)
+	endKey := MakeDataKey(name, "" /* source */, r, endNanos).PrefixEnd()
+	scan := client.Scan(startKey, endKey, 0)
+	if err := db.kv.Run(scan); err != nil {
+		return nil, err
+	}
+	scanResponse := scan.Reply.(*proto.ScanResponse)
+
+	// Construct a new dataSpan for each distinct source encountered in the
+	// query. Each dataspan will contain all data queried from the same source.
+	sourceSpans := make(map[string]*dataSpan)
+	for _, row := range scanResponse.Rows {
+		_, source, _, _ := DecodeDataKey(row.Key)
+		data, err := proto.InternalTimeSeriesDataFromValue(&row.Value)
+		if err != nil {
+			return nil, err
+		}
+
+		if _, ok := sourceSpans[source]; !ok {
+			sourceSpans[source] = &dataSpan{
+				startNanos:  startNanos,
+				sampleNanos: data.SampleDurationNanos,
+				datas:       make([]calibratedData, 0, 1),
+			}
+		}
+		if err := sourceSpans[source].addData(data); err != nil {
+			return nil, err
+		}
+	}
+
+	response := &proto.TimeSeriesQueryResult{
+		Name:    name,
+		Sources: make([]string, 0, len(sourceSpans)),
+	}
+
+	// Create an interpolatingIterator for each dataSpan.
+	iters := make(unionIterator, 0, len(sourceSpans))
+	for name, span := range sourceSpans {
+		response.Sources = append(response.Sources, name)
+		iters = append(iters, span.newIterator())
+	}
+
+	// Iterate through all values in the iteratorSet, adding a datapoint to
+	// the response for each value.
+	iters.init()
+	for iters.isValid() {
+		response.Datapoints = append(response.Datapoints, &proto.TimeSeriesDatapoint{
+			TimestampNanos: iters.timestamp(),
+			Value:          iters.sumAvg(),
```

Comment:
Wondering if `Datapoints` is the desired output format. In particular, the `unionIterators` comments say that you only output a datapoint if one if the `interpolatingIterators` had a measured value at that time. But I would think you'd want to sample the time series at regular `Resolution` intervals. That is, the output should be the timestamp of the first bit of sample data, the size of each interval and then a `repeated Values float64`. You've been thinking about this time series stuff a lot more than I have, though. Feel free to explain why the variable sampling is desirable.